Implement file-based catalog (FBC) support for run bundle

rashmigottipati opened the pull request:
**Description of the change:**

- Create a file-based catalog (FBC) on the fly by generating the bundle, package and channel blobs, if an index image is not provided by the user (i.e. for a default index image `quay.io/operator-framework/opm:latest`) and serve it over a gRPC port.
- If an index image is provided through the flag, then bundle image will be inserted to the index by appending the bundle contents (rendered bundle image) to the index that's converted to a declarative config and this index containing the bundle will be served over a gRPC port.
- If bundle already exists in the index, then we will error out and not proceed with serving the index image.
- Generated FBC will be validated (equivalent to the `opm validate` command) and converted to a string format and sent down to registry pod creation to be served.
- Infer the label type on the index image and indicate there's an FBC label through registry pod for container creation. 
- Added a new container creation command for images that have FBC label on them. For images without the label, old commands (`opm registry add` and `opm registry serve`) that use SQLite under the hood will be called. So support for both (FBC/SQLite) exist at the moment. 

**Motivation for the change:**
https://github.com/operator-framework/operator-sdk/issues/5593 

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

Signed-off-by: rashmigottipati <chowdary.grashmi@gmail.com>

everettraven commented:
I took a look and everything looks good to me. Due to lack of experience in the FBC & OLM space I will leave approval to someone else.

rashmigottipati commented:
/cc @joelanford @jmrodri

VenkatRamaraju commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+
+	if len(bundleDeclConfig.Bundles) > 0 && len(bundleDeclConfig.Channels) > 0 && len(bundleDeclConfig.Packages) > 0 {
+		imageDeclConfig.Bundles = append(imageDeclConfig.Bundles, bundleDeclConfig.Bundles[0])
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
```
I'm trying to think of various possible cases for these two statements. 

For example, if we are attempting to insert a memcached bundle into a "stable" channel of an index image that already has the memcached bundle in an "alpha" channel, we wouldn't want to append the package blob since it will already be present. We'd only want to append the channel and bundle blobs. 

Do we need another check to see if a package already exists?

VenkatRamaraju replied:
@joelanford thoughts on the above scenario?

joelanford replied:
If the bundle already exists in a different channel, then you theoretically already have the package and bundle blob in the FBC. All you have to do at that point is reference it in whatever channel(s) you want it to be in.

joelanford replied:
I'm getting a little bit tripped up by the length checks for the various `bundleDeclConfig` Bundles, Channels, and Packages. Based on the code paths, it seems like a guarantee that those will each contain _exactly_ one package, channel, and bundle in that FBC.

If that's the case, is it possible to refactor things such that the `bundleDeclConfig` length checks can be removed?

Perhaps this function signature could accept individual `pkg declcfg.Package, ch declcfg.Channel, b declcfg.Bundle` parameters? Or a custom unexported struct that allows exactly one of each of those?

I think doing something along those lines would help make it more obvious what this function is doing.

VenkatRamaraju replied:
@joelanford, we initially had it without the checks, but wanted to eliminate _any_ possibility of an index out of bound exception for the array (i.e., `Bundles[0]`). But you're right, with the way the `bundleDeclConfig` object is constructed, this should not happen at all. @rashmigottipati, I'm okay with removing the length checks.

rashmigottipati replied:
@joelanford @VenkatRamaraju I like the idea of having a custom unexported struct that allows exactly one of each. I'll refactor it and get rid of the length checks.

rashmigottipati replied:
@joelanford @VenkatRamaraju Created an unexported struct and got rid of the length checks in [this commit](https://github.com/operator-framework/operator-sdk/pull/5671/commits/a73c3db7cf7bc92e279c6b1c1e446846e1e67944) based on suggestions.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -87,6 +109,71 @@ func (i *Install) setup(ctx context.Context) error {
+		// FBC variables
+		f := &FBCContext{
+			Package:     labels[registrybundle.PackageLabel],
+			ChannelName: strings.Split(labels[registrybundle.ChannelsLabel], ",")[0],
```
Going forward, bundles won't have to have channel metadata included. I think this makes sense _if_ the length of `strings.Split(labels[registrybundle.ChannelsLabel], ",")` is greater than 0. But we should handle a bundle that doesn't define channels as well. 

Question is: what channel name should we fall back to? Auto-generated with a random suffix? Some hardcoded name? The name of the CSV?

Any thoughts or considerations for any of those?

VenkatRamaraju replied:
I think anything prefixed by the operator name is a good fallback, ex: memcached-operator-alpha, nginx-operator-stable, etc.

joelanford replied:
Yeah, that's another option. 

For options that use a non-random name: Would there be any concern with channel name conflicts if you run `operator-sdk run bundle` multiple times against different bundles in the same package?

For options that use a random name: Would there be any concern with the `operator-sdk run bundle` + `operator-sdk run bundle-upgrade` case?

It does seem like a non-random name as a fallback should work though. After all, if the bundle _does_ have a list of channels, we have been using something from that list in the existing `operator-sdk run bundle` code for sqlite-based indexes.

VenkatRamaraju replied:
```
{
    "schema": "olm.channel",
    "name": "alpha",
    "package": "nginx-operator",
    "entries": [
        {
            "name": "nginx-operator.v0.0.1"
        }
    ]
}
```

1. Non-random names - Is running run bundle successively for two different versions of the same operator supposed to be a use case? Each run bundle invocation creates a catalog source. The second time we invoke run bundle, the catalog source object of that operator will already exist, and it'll throw an error. Correct me if I understood your question incorrectly. 

2. Random names - Every channel is uniquely identified by both the name of the channel and its corresponding package.
So even if the channel.name isn't unique, channel.package _will_ be, and so I don't think there's should be any conflict. 

I'm leaning towards non-random names with a meaningful prefix. Anything related to auto generating a random string makes me slightly concerned about randomly generating the same string, even if it's unlikely.

camilamacedo86 replied:
Hi @joelanford, @rashmigottipati @VenkatRamaraju 

I do **not** understand at all why we allow someone to pass the index via the run bundle option _(what is the point? Why/how/where a pre-existent index content maters and is useful data for the purpose of this feature?_ ). 

Assuming that I am right here and we never use the data from index informed ( the only diff is that we add the bundle to index informed instead of creating a new one when it is informed). I'd say why the channel matters? Could we not create any random/static name for all cases since the goal is only to check if the bundle can "run" with OLM?

camilamacedo86 replied:
Btw, could we not remove this option from the run bundle and simplify it

camilamacedo86 replied:
If the index content matter then, should not the channel also matters? If yes, how use a random or static value could address the need? How author would configure the channel if we are ignoring the annotations?

rashmigottipati replied:
Added a check to see if the channel metadata exists on the bundle. If not, setting it a hardcoded value. 
Fixed in [commit](https://github.com/operator-framework/operator-sdk/pull/5671/commits/20f18a072d3a82622896f88897ebf27253ec5284).

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+	// init packages.
+	init := action.Init{
+		Package:        f.Package,
+		DefaultChannel: f.ChannelName,
```
I actually would suggest _not_ doing that. Whatever channel name we decide to generate, that's what we should use for the default channel, I think.

The default channel metadata on the bundle is somewhat of a hack that allows `opm registry add` to modify the package's default channel based on the value of that label (and a few other somewhat magic heuristics like "favor the default channel from the highest semver bundle that defines it")

VenkatRamaraju replied:
I see. In that case, we can leave this as is. The generated channel name will become the default, then.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+	log.SetOutput(os.Stdout)
+
+	// Ensuring a valid bundle size.
+	if len(declcfg.Bundles) < 0 {
```
Are multiple bundles allowed? Should we just check to make sure we have exactly one bundle?

VenkatRamaraju replied:
Yup, exactly one is a better assertion. The `declcfg` is an FBC representation of a single bundle image, and should only have one bundle blob in it.

rashmigottipati replied:
Makes sense. It's exactly one bundle blob, fixed it.

VenkatRamaraju commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+	log.SetOutput(os.Stdout)
+
+	if len(bundleDeclConfig.Bundles) < 0 {
+		return nil, fmt.Errorf("bundles rendered are less than 0 for the bundle image")
```
We can change this to == 1

joelanford replied:
`!= 1`, right?

VenkatRamaraju replied:
Yeah, my bad.

joelanford commented on go.mod:
```diff
@@ -103,6 +103,9 @@ require (
 	github.com/garyburd/redigo v1.6.0 // indirect
 	github.com/ghodss/yaml v1.0.0 // indirect
 	github.com/go-errors/errors v1.0.1 // indirect
+	github.com/go-git/gcfg v1.5.0 // indirect
```
What caused all the changes in `go.mod`? It looks like all of them are `indirect` dependencies, but none of the direct dependencies seem to have changed.

rashmigottipati replied:
Not sure what caused this. I tried removing it and that caused tests to fail stating that I should run `go mod tidy`. I ran `go mod tidy` which added this dependency back.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -64,6 +85,7 @@ func (i Install) Run(ctx context.Context) (*v1alpha1.ClusterServiceVersion, erro
 }
 
 func (i *Install) setup(ctx context.Context) error {
+	var content string
```
Better yet, it seems like we could drop this line if we change line 162 to be `content, err := ...` (with the colon style initialization)

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -87,6 +109,71 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+
+	// generate an fbc if an fbc specific label is found on the image or for a default index image.
+	if i.IndexImageCatalogCreator.HasFBCLabel {
```
Is there a way to restructure this if statement such that we can outdent its body? I wonder if the body should be in its own function?

rashmigottipati replied:
Good idea. Fixed in [commit](https://github.com/operator-framework/operator-sdk/pull/5671/commits/20f18a072d3a82622896f88897ebf27253ec5284)

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -87,6 +109,71 @@ func (i *Install) setup(ctx context.Context) error {
+		// generate a file-based catalog representation of the bundle image
+		declcfg, err = f.createFBC(ctx)
+		if err != nil {
+			log.Errorf("Error in generating file-based catalog with image %s: %v", i.BundleImage, err)
```
If these errors occur, do we end up logging them twice? I wonder if we can return an error with the extra context you've included in the log message. Something like:
```go
return fmt.Errorf("error generating file-based catalog with image %q: %v", i.BundleImage, err)
```

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+
+	// init packages.
+	init := action.Init{
+		Package:        f.Package,
+		DefaultChannel: f.ChannelName,
+	}
+
+	// generate packages.
+	declcfgpackage, err = init.Run()
```
Is it overkill to `action.Init` in this case? I wonder if we can just create the `declcfg.Package` directly since we're just setting a couple of simple string fields.

rashmigottipati replied:
Yup, definitely an overkill to use `action.Init` here. Modified it to directly set those fields similar to generating the channel blob.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+		return nil, fmt.Errorf("bundles rendered are less than 0 for the bundle image")
+	}
+
+	// check if the bundle exists in the index image.
+	if len(bundleDeclConfig.Channels) > 0 && len(bundleDeclConfig.Bundles) > 0 {
+		for _, channel := range imageDeclConfig.Channels {
+			// find the specific channel that the bundle needs to be inserted into.
+			if channel.Name == bundleDeclConfig.Channels[0].Name && channel.Package == bundleDeclConfig.Channels[0].Package {
+				// check if the CSV name is already present in the channel's entries.
+				for _, entry := range channel.Entries {
+					if entry.Name == bundleDeclConfig.Bundles[0].Name {
+						return nil, fmt.Errorf("bundle %q already present in the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+					}
+				}
+
+				break // we only want to search through the specific channel.
+			}
+		}
+	}
```
Can this be simplified to just check `olm.bundle` blobs in the index image? I feel like if the index image contains the bundle _anywhere_ we should fail out, but its possible I'm missing some context for why we're only checking in the specific channel.

rashmigottipati replied:
From my understanding, we should fail out only when the bundle that we are trying to install already exists in the index image, which is consistent with our existing implementation with the sqlite index images where the deprecated `opm registry add` command fails if the bundle we intend to install is present on the index. An index image can have the `olm.bundle` blob with for ex: `memcached-operator.v0.0.1` but we should still be able to install `memcached-operator.v0.0.2` with `run bundle` right?

The reason we are iterating channels is that there can be multiple bundle versions under the same package name so instead we need to find whether the specific version of the bundle exists on the image. So for a matching channel in both bundle image declcfg and index image declcfg, we look for the specific csv name and if it exists, then we error out. Isn't this necessary to do?

VenkatRamaraju replied:
@joelanford I think the reason we didn't do something like

```
for _, bundle := range imageDeclConfig.Bundles {
   if bundle.Name == bundleDeclConfig.Bundles[0].Name && bundle.Package == bundleDeclConfig.Bundles[0].Package {
	// return error
   }
}
```

is because of the case where the bundle may exist, but in a different channel. Example: The index image has the memcached bundle in the "alpha" channel, but we want to the bundle to also be in the "stable" channel (which we are yet to create). The above snippet wouldn't catch that. 

@rashmigottipati for the above example, we'd only need to add a new channel (with this entry) and NOT append the package and bundle blobs, since those already exist (as @joelanford mentioned in [this](https://github.com/operator-framework/operator-sdk/pull/5671#discussion_r853416273) comment). I'm leaning towards having **individual checks** to see if packages, channels, and bundles exist in an index image, and appending accordingly. Curious as to what you think of this approach.

camilamacedo86 replied:
Why operator-sdk run bundle allow us to inform an index?
What is the use? I mean, I understand that the operator-sdk run bundle will install the bundle and let us check the Operator installed via the bundle and using OLM.

So, what value do we bring by allowing a user to provide an index?
What are the use case scenarios? What SDK run bundle can do differ according to the index informed?

Shows me that the only extra test/check that we do here when the index is informed is to verify it would be possible to add the new bundle to a pre-existent index. Am I right? If yes, should it really be part of the run bundle scope and checks? Should not it only be used/useful for the upgrade bundle command?

I added it in : https://github.com/operator-framework/enhancements/pull/112/files#r856161326 as well

rashmigottipati replied:
@joelanford @VenkatRamaraju Modified the validation logic to insert the `bundle` into `index` only when `package`, `channel` and `bundle` blobs don't exist in the index. There's a case where a `bundle` already exists in a different channel and we would now like to add it to another channel, in which case, both bundle and package blobs will exist on the FBC and we need to add the bundle's channel to the index. 

Fixed in [commit](https://github.com/operator-framework/operator-sdk/pull/5671/commits/99f9c4c6c11d63cc78c1c2a3064227b489fa83c5).

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
+
+		if len(bundleDeclConfig.Others) > 0 {
+			imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others[0])
+		}
```
```suggestion
		imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others...)
```

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+
+// createFBC generates an FBC by creating bundle, package and channel blobs.
+func (f *FBCContext) createFBC(ctx context.Context) (*declarativeconfig.DeclarativeConfig, error) {
+	var (
+		declcfg        *declarativeconfig.DeclarativeConfig
+		declcfgpackage *declarativeconfig.Package
+		err            error
+	)
```
Is it possible to remove these declarations and use `:=` style initializations instead?

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+
+	declcfg.Channels = []declarativeconfig.Channel{channel}
+
+	return declcfg, nil
```
Going along with my previous question about the length checks in `addBundleToIndexImage`, perhaps we could declare a new unexported struct like the following that has exactly one of the things we know we're generating:

```go
type bundleDeclcfg struct {
	Package declcfg.Package
	Channel declcfg.Channel
	Bundle declcfg.Bundle
}
```
And then use that for the `declcfg` variable in this function in place of the more lenient `declcfg.DeclarativeConfig` type. I think that would remove the need for the length checks in `addBundleToIndexImage`.

The only catch is that there would need to be some (admittedly very simple) code to convert that to a `declcfg.DeclarativeConfig` struct in the case where we don't have an existing index image.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+// validateFBC converts the generated declarative config to a model and validates it.
+func validateFBC(declcfg *declarativeconfig.DeclarativeConfig) error {
+	// convert declarative config to model
+	FBCmodel, err := declarativeconfig.ConvertToModel(*declcfg)
```
```suggestion
	fbcModel, err := declarativeconfig.ConvertToModel(*declcfg)
```

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+		return err
+	}
+
+	if err = FBCmodel.Validate(); err != nil {
+		log.Errorf("Error validating the generated FBC: %v", err)
+		return err
+	}
```
There's no need to call `Validate` after the conversion ([`ConvertToModel` validates internally](https://github.com/operator-framework/operator-registry/blob/9e59597a435341d857283011c8ede1c16e5e292b/alpha/declcfg/declcfg_to_model.go#L167-L169)).

rashmigottipati replied:
Didn't realize it was validating it internally. Got rid of the `Validate()` function.

joelanford commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -87,6 +89,12 @@ type RegistryPod struct { //nolint:maligned
 
 	cfg *operator.Configuration
+
+	// FBC data
+	FBCcontent  string
+	FBCdir      string
+	FBCfile     string
+	HasFBCLabel bool
```
With this change, is there still a need for the `BundleItems` field for backwards compatibility?

It seems like we're perhaps mixing sqlite-specific registry pod concerns with FBC registry pod concerns?

I wonder if we should have separate structs for `SqliteRegistryPod` vs `FBCRegistryPod`?

joelanford replied:
Also, some Go idiomatic naming nits:
```suggestion
	// FBC data
	FBCContent  string
	FBCDir      string
	FBCFile     string
	HasFBCLabel bool
```

rashmigottipati replied:
@joelanford I split RegistryPod into two different structs (`SQLiteRegistryPod` and `FBCRegistryPod`), so the code paths for these remain separate (there's `registry_pod.go` file that handles the sqlite stuff and `fbc_registry_pod.go` file in a different package for all of the FBC stuff). 

When we deprecate SQLite stuff and completely remove the commands, the `registry_pod.go` file can go away. Let me know if this is the right way go about it or if we should refactor it in a different way. I'm open to your suggestions.

camilamacedo86 commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -313,20 +329,29 @@ opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMo
+	if rp.HasFBCLabel {
+		t = template.Must(template.New("cmd").Funcs(funcMap).Parse(fbcCmdTemplate))
+	} else {
+		t = template.Must(template.New("cmd").Funcs(funcMap).Parse(cmdTemplate))
```
Why not instead of have 2 commands to keep maintained we do not change it for:

```go
if isSQLIndex { 
      // convert to FBC index 
}
```
So that no matter the format we do the same steps and work on in the same.

VenkatRamaraju replied:
@camilamacedo86,

Based on some of the discussions we had recently with @joelanford, rendering SQLite index images to FBCs in SDK code is not a viable approach due to cross-compilation conflicts with Operator Registry code.

rashmigottipati replied:
@camilamacedo86 

Adding [comment](https://github.com/operator-framework/enhancements/pull/112#discussion_r856412131) from the EP for your reference

VenkatRamaraju commented on internal/olm/operator/bundle/install.go:
```diff
@@ -87,6 +109,65 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+
+	// set the field to true if FBC label is on the image or for a default index image.
+	if _, hasFBCLabel := catalogLabels[containertools.ConfigsLocationLabel]; hasFBCLabel || i.IndexImageCatalogCreator.IndexImage == registry.DefaultIndexImage {
+		i.IndexImageCatalogCreator.HasFBCLabel = true
+	}
```
@rashmigottipati I just realized that we are not actually outputting any logs related to the deprecation of the SQLite index. All the opm-related warnings get written out to the registry pod logs, which may/may not be seen by the operator author. It might be valuable to add a warning statement as an else block here saying "The use of SQLite index images is deprecated and will be completely removed in a future release, please migrate your catalogs to the new File-Based Catalog format" so that a user is notified of the transition every time run bundle is invoked with a SQLite index image.

rashmigottipati replied:
That's right. I realized that opm related warnings get written out to registry pod logs instead. Added a deprecation log for SQLite index images in my latest commit.

VenkatRamaraju commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -87,6 +89,12 @@ type RegistryPod struct { //nolint:maligned
 	cfg *operator.Configuration
+
+	// FBC data
+	FBCContent  string
+	FBCDir      string
+	FBCFile     string
+	HasFBCLabel bool
```
We should describe each one of these with a comment similar to all the other fields.

rashmigottipati commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -99,6 +107,14 @@ func (rp *RegistryPod) init(cfg *operator.Configuration) error {
 
+	// if the FBC label is set to true, assign FBCdir and FBCfile variables to serve the content from.
+	if rp.HasFBCLabel {
+		bundleImage := rp.BundleItems[len(rp.BundleItems)-1].ImageTag
```
There a validate function that gets called right before making this assignment. That function ensures that the bundle items has atleast one item in it.

VenkatRamaraju commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+		return "", fmt.Errorf("error creating a File-Based Catalog with image %q: %v", bundleImage, err)
+	}
+
+	declcfg := &declarativeconfig.DeclarativeConfig{
+		Bundles:  []declarativeconfig.Bundle{bundleDeclcfg.Bundle},
+		Packages: []declarativeconfig.Package{bundleDeclcfg.Package},
+		Channels: []declarativeconfig.Channel{bundleDeclcfg.Channel},
+	}
```
I may have missed this in the review. Why are we creating a new object with the same values as the original `bundleDeclcfg`?

VenkatRamaraju replied:
Coming back to this. We created the unexported struct to remove all checks of != 1. Isn't that what needs to be used here?

`createFBC` will return a "minimal" FBC (with one in each field), and so we'd create the `bundleDeclcfg` object here and propagate it further, correct?

VenkatRamaraju replied:
Disregard, I just saw that the createFBC return type changed since I last reviewed it. This is fine.

rashmigottipati replied:
@VenkatRamaraju `createFBC` will return a minimal FBC that is of the `bundleDeclcfg` type. I'm creating `declcfg` to convert the `bundleDeclcfg` to the more `declarativeconfig.DeclarativeConfig` type as `addBundleToIndexImage` function expects that type. 

I think the idea here was to create this unexported struct to get rid of the length checks. We are creating exactly one element of each type i.e bundles/channels/packages during `createFBC` invocation and then we need to convert this to the `declarativeconfig.DeclarativeConfig` type before `addBundleToIndexImage` so we know that it has exactly one element of each of those.

rashmigottipati commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -0,0 +1,250 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fbcindex
```
@VenkatRamaraju that's right. So I think the intent is to keep the two structs separate and have the code paths separate, so we can remove the SQLite code path when it's deprecated and completely removed. 
So these two structs have different fields (some are common), and the initialization and container creation template are different. That's about it.

VenkatRamaraju reviewed (approved):
/lgtm

everettraven reviewed (approved):
LGTM

rashmigottipati commented:
@joelanford @jmrodri can you please review this PR again? Addressed all of the review feedback and looking to get your approvals too before I merge it.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+}
+
+// addBundleToIndexImage adds the bundle to an existing index image if the bundle is not already present in the index image.
+func addBundleToIndexImage(ctx context.Context, indexImage string, bundleDeclConfig *declarativeconfig.DeclarativeConfig) (*declarativeconfig.DeclarativeConfig, error) {
```
Should this function accept `bundleDeclcfg` instead of `declarativeconfig.DeclarativeConfig`? Seems like we're assuming we have exactly one each of package/channel/bundle.

rashmigottipati replied:
It's probably better to leave it as is, because we are calling `validateFBC` function  right after this call and that function expects the type `declarativeconfig.DeclarativeConfig` to convert to a model. So this conversion from `bundleDeclcfg` to `declarativeconfig.DeclarativeConfig` needs to happen anyways. WDYT?

rashmigottipati replied:
Ignore my previous comment. :)

I've [modified it](https://github.com/operator-framework/operator-sdk/pull/5671/commits/21dc19831a1a23a0df944252514f5d5449accb08) based on your suggestion and now converting to the `DeclarativeConfig` type just before the validation checks. I think this makes sense as the intent is to have the function expect a bundle declarative config with exactly one of bundles/packages/channels.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+
+	log.Infof("Inserted the new bundle %q into the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+
```
I _think_ we can significantly simplify this insertion logic.

First, we can iterate `imageDeclConfig.Bundles`. If we find a bundle that matches the package name and bundle name of `bundleDeclConfig`, we stop and error out with "bundle already exists in the catalog image".

If we get through that loop without erroring, we can be sure that the bundle isn't in the catalog, so then we know we're adding at least the bundle. Then we have the separate question of whether we need to add/update the package and channel blob associated with that bundle.

It seems like the existing logic actually permits `run bundle` to succeed even if the catalog image already contains the bundle (i'm thinking lines 258-262), which I think we said should not be allowed.

VenkatRamaraju replied:
The current logic will succeed if the bundle is present but the channel is _not_ present. This is a use case we want to support, right? For any other case where the bundle exists, the current logic will error out, as we want it to. If we don't want to support the use case above, then I agree with your logic. 

You mentioned **updating** the package/channel. In what scenario would be be _updating_ anything in the FBC? It seems like the only thing we need to do throughout this run bundle call is add items to generate an FBC, and not really modify any package/channel inside. Or have I misunderstood the use of that word in this context?

rashmigottipati replied:
@joelanford could you please clarify whether we need to support the use-case of updating the channel/package if the bundle is already present. 

The reason why we are permitting `run bundle` to succeed is that a bundle can exist in the index image but in a different channel and not in the channel we want to add it in - in this case, the bundle would already exist in the index but not the channel, so we need to append the channel from `bundleDeclConfig` to `imageDeclConfig`. Shouldn't we be doing this? 

I made this change based on your earlier review comment [here](https://github.com/operator-framework/operator-sdk/pull/5671#discussion_r853416273) although I may have misunderstood it, so looking to get these clarifications from you on whether we could keep this existing logic or change it. Please let me know.

rashmigottipati replied:
@joelanford Any thoughts on this? Can we proceed with the assumption that we need to update the channel if the bundle is already present in the index?

joelanford replied:
In sqlite, if you try to `opm add` a bundle that's already in the DB, regardless of which channels its in, `opm add` will fail. So `operator-sdk run bundle` inherits that behavior when dealing with sqlite catalogs. 

I think we should have equivalent logic for FBC -- basically if the bundle exists in the catalog, we'll fail no matter the channel differences.

rashmigottipati replied:
@joelanford I agree. That makes sense. I've simplified the insertion logic to fail if bundle is present on index and add all three (channel/package/bundle) if none of them exist on the index.

joelanford replied:
If the bundle's package is already present, but the bundle is not, it seems like this logic causes a failure in that case?

It seems like once we've determined that the bundle is not present in the catalog, we should definitely be adding the bundle.

This [comment](https://github.com/operator-framework/enhancements/pull/112#discussion_r866025776) sums up what it seems like this logic needs to be:

> 1.  Does this bundle name/package already exist in the catalog? If so, fail.
> 2.  Does this bundle's package already exist in the catalog? If so, skip package blob generation
> 3.  Generate a channel blob using a name that does not conflict with an existing channel in the bundle's package. Maybe hardcoded as operator-sdk-run-bundle?
> 4.  Generate the bundle blob.
> 5. Insert the generated FBC into <fbcRootDir>/operator-sdk-run-bundle/extra.yaml and run the catalog image.

Since we aren't yet doing the "only add the extra FBC" stuff yet (but we plan to), I think the logic should fail out if the generated channel name conflicts with an existing channel name. And I think we should pick a channel name that is exceedingly unlikely to conflict (e.g. `operator-sdk-run` or something along those lines)

So, summing up:
1. If the rendered bundle is already present -- fail
2. If the generated channel is already present -- fail
3. If the generated package is already present - skip package generation
4. Add the bundle and channel -- and optionally the package (based on step 3).

VenkatRamaraju replied:
@joelanford, the latest commit reflects this logic, PTAL.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+	// Rendering the bundle image into a declarative config format.
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: f.Refs,
```
Do we expect `Refs` to always be length 1? If so, should we change the `FBCContext` struct to have a singular `Ref`?

And do we expect that ref to be a bundle? If so, let's set `action.Render`'s `AllowedRefMask`. You can check for [`ErrNotAllowed`](https://github.com/operator-framework/operator-registry/blob/aa8db09278a0ea30dd7cf07e3b197a3d7b5ce034/alpha/action/render.go#L49) when `Run()` returns and have your own error message along the lines of "only bundle images can be specified for `run bundle`"

VenkatRamaraju replied:
I'd prefer if we keep FBCContext as a string array. In some of the run bundle-upgrade work, we use this refs field for the previous injected bundles so that we can render all of them together.

rashmigottipati replied:
As @VenkatRamaraju mentioned, we are using the refs field to render both index image and bundle image, so the len varies for upgrade functionality.

joelanford commented on internal/olm/operator/bundle/install.go:
```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+// stringifyDecConfig writes the generated declarative config to a string.
+func stringifyDecConfig(declcfg *declarativeconfig.DeclarativeConfig) (string, error) {
+	var buf bytes.Buffer
+	err := declarativeconfig.WriteJSON(*declcfg, &buf)
```
Maybe consider `YAML` instead? I think YAML is generally smaller than JSON. So we'd be less likely to run into the etcd size limit issues I mentioned before.

rashmigottipati replied:
Sounds good, I think that will partially fix the size limit issues. I'll change it to encode to YAML instead.

VenkatRamaraju commented on internal/olm/operator/bundle/install.go:
```diff
@@ -87,6 +117,46 @@ func (i *Install) setup(ctx context.Context) error {
 	}
 
+	// get index image labels.
+	catalogLabels, err := registryutil.GetImageLabels(ctx, nil, i.IndexImageCatalogCreator.IndexImage, false)
+	if err != nil {
+		return fmt.Errorf("get index image labels: %v", err)
+	}
+
+	// set the field to true if FBC label is on the image or for a default index image.
+	if _, hasFBCLabel := catalogLabels[containertools.ConfigsLocationLabel]; hasFBCLabel || i.IndexImageCatalogCreator.IndexImage == registry.DefaultIndexImage {
+		i.IndexImageCatalogCreator.HasFBCLabel = true
```
When the index image is the default one, we'd still be making this API call to containertools even though it would have no effect on the if block and codepath. Ideally, we could save _some_ time by avoiding this call.

Maybe we can do this in a future improvement. What do you think?

rashmigottipati replied:
I guess the intention for this is that for the default index image, we want to default to FBC creation and not the old sqlite route. I'm assuming you're referring to the API calls to Operator Registry.

VenkatRamaraju replied:
`containertools.ConfigsLocationLabel` on line 127 is an API call, correct? However, if the index image is the default one, the `_, hasFBCLabel := catalogLabels[containertools.ConfigsLocationLabel]; hasFBCLabel` part of the if does not matter since we know we want to use FBC for the default index image. However, even though it does not matter, the API call is being made in the current implementation. I was thinking we wouldn't need to check what the label on the image is if the image is the default index image. This might be as simple as reversing the order of the two conditions that are joined by `||`.

rashmigottipati replied:
I see your point. Although I agree with you, I don't think it's syntactically possible to switch these two conditions as the second condition `_, hasFBCLabel := catalogLabels[containertools.ConfigsLocationLabel]; hasFBCLabel` is a map lookup which is combination of a condition (evaluates `hasFBCLabel` which will be true if key exists in the map) and initialization (which is setting `hasFBCLabel` to true). 

In my opinion, the only way to do this is invoke the API call and set `hasFBCLabel` before this if statement, and we can do something like `if i.IndexImageCatalogCreator.IndexImage == registry.DefaultIndexImage || hasFBCLabel`, but it still wouldn't avoid the API call.

VenkatRamaraju replied:
Alright, we can leave it as is.

VenkatRamaraju reviewed (approved):
One comment that can be addressed at a later date. Apart from that, looks good to me!

/lgtm

jmrodri reviewed (approved):
/lgtm

joelanford reviewed (approved):
/approve
/lgtm

Great job!

I think we'll need to somewhat quickly revisit the implementation related to [this comment](https://github.com/operator-framework/enhancements/pull/112#discussion_r865250033), but that could be done in a follow-up IMO. Do we have an issue to track that work? If not can we make one so we don't lose track?

rashmigottipati commented:
@joelanford thanks a lot for your review! Will create an issue to track the comment about injecting just the generated FBC and address in a follow-up ASAP.
